Feat: Order suite

essteer opened the pull request:
Adds the Order class and related repo, service and controller layers.

Adds unit test classes for entity and service layers. All tests passing.

Additional mappings are saved in draft form on the branch `feat-orders-extended-mappings`.

essteer committed:
feat: Add Order abstract class, enums for order status and type, and OpenOrder class

essteer committed:
feat: Add Order concrete class impl LimitOrder and MarketOrder using base and quote fx terminology

essteer committed:
Merge branch 'develop' into feat-orders

essteer committed:
feat!: Consolidate orders into concrete Order class and remove child classes and OpenOrder class

essteer committed:
feat: Add initial repo and service layers for Order class

essteer committed:
feat: Add methods to find Order by User ID in combination with enum attributes

essteer committed:
feat: Add extended retrieval mappings for Order suite

essteer committed:
refactor: Amend methods in Order suite layers and add unit tests for Order and OrderService classes

essteer committed:
Merge branch 'develop' into feat-orders

essteer committed:
chore: Add CLEARED enum to OrderStatus to reflect fully complete orders and reassign CLOSED for partially filled orders that are closed due to insufficient funds

dexterklui commented on src/main/java/com/fdmgroup/forex/enums/OrderStatus.java:
```diff
@@ -0,0 +1,5 @@
+package com.fdmgroup.forex.enums;
+
+public enum OrderStatus {
+    ACTIVE, CANCELLED, CLEARED, CLOSED, EXPIRED
```
Do you think it's a good idea to add JavaDoc here to explain what each enum mean?
Especially CANCELLED vs CLEARED vs CLOSED

essteer replied:
Yes, will do

essteer replied:
See [593c591](https://github.com/forex-agile/backend/pull/8/commits/593c59162fe1595128ade621afe5f4076d010789)

essteer committed:
chore: Add javadocs to describe enum states for OrderStatus

essteer committed:
feat: Throw IllegalArgumentException for invalid total, residual and limit values, and ensure checks take place via method calls from constructors

essteer committed:
Merging develop branch locally to resolve potential conflicts ahead of PR

essteer committed:
test: Update tests to mock new User and Portfolio behaviour

essteer committed:
chore: Remove unnecessary autowired annotation

dexterklui commented on src/main/java/com/fdmgroup/forex/enums/OrderStatus.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Denotes the state of the order as follows:
+ * ACTIVE: Order is live and open to matches
+ * CANCELLED: Order cancelled by the user
```
Order can be CANCELLED on InsufficientFundException too, when no partial trades ever happened

essteer replied:
[a687131](https://github.com/forex-agile/backend/pull/8/commits/a687131dfe403beabca4640ce17d97442725c208)

essteer committed:
docs: Update javadocs for OrderStatus CANCELLED condition

dexterklui committed:
feat: add option to create sample data on start up

To populate the database with sample data if not exists, for development

dexterklui committed:
feat: remove full error log from HTTP response

By handling MethodArgumentTypeMismatchException raised in controller
methods, when HTTP requests contain illegal value for an enum.

This is just a temporary fix. We will need to have custom message in the
response, specifying what is wrong with the HTTP request.

dexterklui committed:
fix(Order): changing a column name in database

Changing "limit" to "order_limit" for the column name in the database.
Because "limit" is a reserved word for MySQL, leading to error.

dexterklui committed:
refactor(Order): use constructor chaining

essteer commented on src/main/java/com/fdmgroup/forex/DataLoader.java:
```diff
@@ -33,4 +63,30 @@ private void createRoles() {
+	private void createSampleData() {
+		Role role = roleRepo.findByRole(defaultRole).get();
+
+		Currency hkd = currencyRepo.findById(usdCode).orElseGet(() -> {
```
This should be `hkd = currencyRepo.findById(hkdCode)` -> currently reads `usdCode`

essteer reviewed (commented):
@dexterklui This looks fine to me - just the one fix as per the comment. I can't mark this as "Request changes" for some reason

dexterklui committed:
fix(DataLoader): passing wrong variable name

dexterklui committed:
feat: better response message for bad request

Continuing from 2f9e15b, now the response message on
MethodArgumentTypeMismatchException will say what path variable has
invalid value, and if the type is an enum, the response will list valid
values.

dexterklui committed:
refactor: forgot to use extracted variable

dexterklui committed:
feat(Order): make user field not updatable

dexterklui committed:
refactor(Order): remove @inheritance annotation

Because we are not having inheritance for Order entity

dexterklui committed:
fix(Order): should allow updating order status

dexterklui committed:
fix(Order): misleading error message
